File: 543_Diameter_of_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.maximum = 0
        def traverse(root):
            if root is None:
                return 0
            left = traverse(root.left)
            right = traverse(root.right)
            self.maximum = max(self.maximum, left + right)
            return max(left, right) + 1
        traverse(root)
        return self.maximum

File: 543_Diameter_of_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int maximum = 0;
        traverse(root, maximum);
        return maximum;
    }
private:
    int traverse(TreeNode* node, int& maximum) {
        if (!node) return 0;
        int left = traverse(node->left, maximum);
        int right = traverse(node->right, maximum);
        maximum = max(maximum, left + right);
        return 1 + max(left, right);
    }
};
